beluga_amcl/include/beluga_amcl/ros2_common.hpp

```diff
@@ -43,369 +47,64 @@ constexpr std::string_view kNav2DifferentialModelName = "nav2_amcl::Differential
-    descriptor.description = "Execution policy used to process particles [seq, par].";
-    node.declare_parameter("execution_policy", "seq", descriptor);
-  }
-
-  {
-    auto descriptor = rcl_interfaces::msg::ParameterDescriptor();
-    descriptor.description = "Whether the node should configure and activate itself or not.";
-    node.declare_parameter("autostart", false, descriptor);
-  }
-
-  {
-    auto descriptor = rcl_interfaces::msg::ParameterDescriptor();
-    descriptor.description = "Delay, in seconds, before autostarting if autostarting.";
-    node.declare_parameter("autostart_delay", 0.0, descriptor);
-  }
-}
-
+/// Base AMCL lifecycle node, with some basic common functionalities, such as transform tree utilities, common
+/// publishers, subscribers, lifecycle related callbacks and configuration points, enabling extension by inheritance.
+class BaseAMCLNode : public rclcpp_lifecycle::LifecycleNode {
+ public:
+  /// Constructor, following LifecycleNode signature.
+  explicit BaseAMCLNode(
+      const std::string& node_name = "amcl",
+      const std::string& node_namespace = "",
+      const rclcpp::NodeOptions& node_options = rclcpp::NodeOptions{});
+
+  /// Destructor for the base AMCL node.
+  ~BaseAMCLNode() override;
+
+ protected:
+  /// Callback for lifecycle transitions from the UNCONFIGURED state to the INACTIVE state.
+  CallbackReturn on_configure(const rclcpp_lifecycle::State&) override;
+
+  /// Callback for lifecycle transitions from the INACTIVE state to the ACTIVE state.
+  CallbackReturn on_activate(const rclcpp_lifecycle::State&) override;
+
+  /// Callback for lifecycle transitions from the ACTIVE state to the INACTIVE state.
+  CallbackReturn on_deactivate(const rclcpp_lifecycle::State&) override;
+
+  /// Callback for lifecycle transitions from the INACTIVE state to the UNCONFIGURED state.
+  CallbackReturn on_cleanup(const rclcpp_lifecycle::State&) override;
+
+  /// Callback for lifecycle transitions from most states to the FINALIZED state.
+  CallbackReturn on_shutdown(const rclcpp_lifecycle::State&) override;
+
+  // Configuration points for extra steps needed for the different AMCL variants.
+
+  /// Extra steps for the on_configure callback. Defaults to no-op.
+  virtual void on_configure_extra([[maybe_unused]] const rclcpp_lifecycle::State& state) {}
+  /// Extra steps for the on_deactivate callback. Defaults to no-op.
+  virtual void on_deactivate_extra([[maybe_unused]] const rclcpp_lifecycle::State& state) {}
+  /// Extra steps for the on_shutdown callback. Defaults to no-op.
+  virtual void on_shutdown_extra([[maybe_unused]] const rclcpp_lifecycle::State& state) {}
+  /// Extra steps for the on_cleanup callback. Defaults to no-op.
+  virtual void on_cleanup_extra([[maybe_unused]] const rclcpp_lifecycle::State& state) {}
+  /// Extra steps for the on_activate callback. Defaults to no-op.
+  virtual void on_activate_extra([[maybe_unused]] const rclcpp_lifecycle::State& state) {}
```

Comment:
@serraramiro1 nit^2: consider `do_<event>` instead of `on_<event>_extra` for naming. It's more common for NVI patterns.

Reply:
Done! 
This PR is failing working with gcc but failing with clang.

Have you ever seen this before?
Invoking @glpuga and @nahueespinosa JIC you happen to have faced this.

Reply:
Done, weird `[[maybe_unused]]` behavior, WTF...